File: cookbooks/from-sql-to-arcflow/02-confidence-tiered-query.py
"""
Question 2: "Show employment relationships, but only the high-confidence ones."

In SQL, confidence is a column on a join table — you filter by it like any
other predicate. In ArcFlow, confidence is a first-class property on the
edge with semantic meaning the engine knows about (algorithms can weight
by it, standing queries can fire on threshold crossings, etc.).

This recipe shows the basic mechanics; the deeper value (confidence-
weighted PageRank, drift detection, calibrated retrieval) is in the
spatiotemporal-tactical-queries cookbook.
"""

from _common import make_duckdb, make_arcflow

print("=" * 68)
print("Q2: High-confidence employment relationships only (confidence > 0.85)")
print("=" * 68)

# ──────────────────────── DuckDB / SQL ────────────────────────
print("\n--- DuckDB / SQL ---")
sql = """
SELECT p.name AS person, o.name AS org, e.confidence
FROM persons p
JOIN employment e ON p.id = e.person_id
JOIN orgs o ON o.id = e.org_id
WHERE e.confidence > 0.85
ORDER BY e.confidence DESC
LIMIT 5
"""
print(sql.strip())
print("\nresult:")
db = make_duckdb()
for r in db.execute(sql).fetchall():
    print(f"  person={r[0]!r:>10} org={r[1]!r:>12} conf={r[2]:.2f}")
db.close()

# ──────────────────────── ArcFlow / Cypher ────────────────────────
print("\n--- ArcFlow / Cypher ---")
cypher = """
MATCH (p:Person)-[e:EMPLOYED_AT]->(o:Org)
WHERE e.confidence > 0.85
RETURN p.name AS person, o.name AS org, e.confidence AS conf
ORDER BY conf DESC
LIMIT 5
"""
print(cypher.strip())
print("\nresult:")
af = make_arcflow()
for r in af.execute(cypher):
    print(f"  person={r['person']!r:>10} org={r['org']!r:>12} conf={float(r['conf']):.2f}")
af.close()

print("\n--- Observation ---")
print("Same shape — both are simple filtered projections. The difference is")
print("downstream: ArcFlow algorithms (algo.confidencePageRank,")
print("algo.confidencePath) consume `_confidence` natively. SQL has no native")
print("graph-algorithmic primitive that knows about confidence.")

File: cookbooks/from-sql-to-arcflow/rust/src/bin/02-confidence-tiered-query.rs
//! Step 02 (Rust) — confidence-tiered query, both engines.
//!
//! Mirrors `02-confidence-tiered-query.py`. The flat filter is similar
//! in both engines; the lift is that ArcFlow's confidence column is
//! also consumed natively by confidence-aware algorithms
//! (algo.confidencePageRank, algo.confidencePath) when the question
//! needs propagation through joins.

use anyhow::Result;
use from_sql_to_arcflow::load_both;

const MIN_CONF: f64 = 0.85;

fn main() -> Result<()> {
    let both = load_both()?;

    println!("=== Employment relationships with confidence >= {MIN_CONF} ===\n");

    println!("--- SQL (DuckDB) ---");
    let sql_q = format!(
        "SELECT p.name AS person, o.name AS org, e.confidence \
         FROM employment e \
         JOIN persons p ON p.id = e.person_id \
         JOIN orgs o ON o.id = e.org_id \
         WHERE e.confidence >= {MIN_CONF} \
         ORDER BY e.confidence DESC",
    );
    println!("{sql_q}");
    let mut stmt = both.sql.prepare(&sql_q)?;
    let rows: Vec<(String, String, f64)> = stmt
        .query_map([], |row| Ok((
            row.get::<_, String>(0)?,
            row.get::<_, String>(1)?,
            row.get::<_, f64>(2)?,
        )))?
        .filter_map(|r| r.ok())
        .collect();
    for (person, org, conf) in &rows {
        println!("  {person:>8} @ {org:<12} conf={conf:.2}");
    }
    println!();

    println!("--- WorldCypher (ArcFlow) ---");
    let cy = format!(
        "MATCH (p:Person)-[r:WORKED_AT]->(o:Org) \
         WHERE r.confidence >= {MIN_CONF} \
         RETURN p.name AS person, o.name AS org, r.confidence AS confidence \
         ORDER BY r.confidence DESC"
    );
    println!("{cy}");
    let result = both.graph.execute(&cy)?;
    for row in &result.rows {
        println!(
            "  {:>8} @ {:<12} conf={}",
            row.get("person").cloned().unwrap_or_default(),
            row.get("org").cloned().unwrap_or_default(),
            row.get("confidence").cloned().unwrap_or_default(),
        );
    }

    println!("\nVerdict: same shape in both engines for the flat filter.");
    println!("Lift in ArcFlow: the same confidence column powers algo.confidencePageRank,");
    println!("algo.confidencePath, and live-proof assertions — no application-layer plumbing.");
    Ok(())
}
